### PR TITLE
Improving documentation of Model

### PR DESCRIPTION
**Pull request by romaninsh**
Currently our Model documentation lacks:

image

As we are adding a lot of things, I'm restructuring the documentation there entirely:

image

Topics covered:

 - [x] Model what it is etc
 - [x] Actions (basics)
 - [ ] Action (in details)
 - [x] Fields (basics)
 - [ ] Fields (in details)
 - [x] Persistence binding
 - [x] Static persistence (basic)
 - [ ] Static Persistence (in details)
 - [x] Validation (basic)
 - [x] Inheritance (basic)
 - [ ] SubTypeSwitch field #417

### CODE REVIEW

File: docs/model.rst [hunk 1]
```diff
@@ -5,21 +5,152 @@
...
+
+   $all_users = new User($db); // John, Susan, Leo, Bill, Karen
+
+Certain operations may "shrink" this set, such as adding conditions::
+
+   $male_users = $all_users->addCondition('gender', 'M');
+
+   send_email_to_users($male_users);
+
+This essentially filters your users without fetching them from the database server. Note that most
```
Comment (PhilippGrashoff) [inline docs/model.rst]: As send_email_to_users would sure fetch the records from database, I find this sentence misleading together with the sample code misleading.
Either drop the database part of the sentence, or, if you wish to explicitly point out that addCondition() does not perform a request, give it an extra sentence? Like `addContition()` does not actually load any data from the database, but all upcoming database requests for $male_users will apply this condition on request.

File: docs/model.rst [hunk 2]
```diff
@@ -31,34 +162,317 @@ persistence driver. Use it to define fields of your model::
...
+
+.. php:method:: addCalculatedField($name, $callback)
+
+Creates new field object inside your model. Field value will be automatically
+calculated by your callback method right after loading data record in model::
+
+   $this->addField('term', ['caption'=>'Repayment term in months', 'default'=>36]);
+   $this->addField('rate', ['caption'=>'APR %', 'default'=>5]);
+
+   $this->addCalculatedField('interest', function($m) use($apr_calculator) {
```
Comment (PhilippGrashoff) [inline docs/model.rst]: To make the sample code consise, I think `$apr_calculator` should be defined in sample code.
Or make code easier without `use()` and just have `return $m->calculateInterest();` ?

File: docs/model.rst [hunk 3]
```diff
@@ -31,34 +162,317 @@ persistence driver. Use it to define fields of your model::
...
-    By default model will only allow you to operate with values for the fields
-    that have been defined through addField(). If you attempt to get, set or
-    otherwise access the value of any other field that has not been properly
-    defined, you'll get exception. Read more about :php:class:`Field`
+By default model will only allow you to operate with values for the fields
+that have been defined through addField(). If you attempt to get, set or
+otherwise access the value of any other field that has not been properly
+defined, you'll get exception. Read more about :php:class:`Field`
+
+If you set strict_field to false, then the check will not be performed.
```
Comment (PhilippGrashoff) [inline docs/model.rst]: strict_field or strict_fields? Headline is strict_fields

**Review by PhilippGrashoff — CHANGES_REQUESTED** [review]
added 3 comments, please check.

File: docs/model.rst (hunk 1 above)
Comment (romaninsh) [reply docs/model.rst]: addressed.

**Review by PhilippGrashoff — APPROVED** [review]
LGTM

File: docs/model.rst [hunk 4]
```diff
@@ -5,21 +5,157 @@
...
+
+         $this->hasMany('Invoices', Invoice::class); 
+      }
+   }
+
+PHP does not have a "class" type, so `Invoice::class` will translate into a string "yourapp\Model\Invoice"
+and is a most efficient way to specify related class name.
+
+You way also use `new Invoice()` there but be sure not to specify any argument, unless you intend
+to use cross-persistence referencing (this is further explained in Advanced section)
```
Comment (DarkSide666) [inline docs/model.rst]: this is very good note!

File: docs/model.rst [hunk 5]
```diff
@@ -31,34 +167,329 @@ persistence driver. Use it to define fields of your model::
         }
     }
 
+You may safely rely on `$this->persistence` property to make choices::
+
+   if ($this->persistence instanceof \atk4\data\Persistence\SQL) {
```
Comment (DarkSide666) [inline docs/model.rst]: also very good and useful example

File: docs/model.rst [hunk 6]
```diff
@@ -31,34 +167,329 @@ persistence driver. Use it to define fields of your model::
...
+
+Other Uses
+^^^^^^^^^^
+
+Other uses for model hooks are explained in :ref:`Hooks`
+
+
+Inheritance
+-----------
+ATK Data models are really good for structuring hierarchically. Here is example::
```
Comment (DarkSide666) [inline docs/model.rst]: This is not that type of _hierarchically_ I thought about.
You're talking about extending or sub-typing model classes below. It's not hierarchical structure i guess. Hierarchical would be parent/child relation + relation of class to itself (tree structure).

File: docs/model.rst [hunk 7]
```diff
@@ -31,34 +167,329 @@ persistence driver. Use it to define fields of your model::
...
+There are some advanced techniques like "SubTypes" or class substitution,
+for example, this hook may be placed in the "User" class init()::
+
+   $this->addHook('afterLoad', function($m) {
+      if ($m['purchases'] > 1000) {
+         $this->breakHook($this->asModel(VIPUser::class);
+      }
+   });
+
+See also :php:class:`Field\SubTypeSwitch`
```
Comment (DarkSide666) [inline docs/model.rst]: Do we have that? What is it???

**Review by DarkSide666 — APPROVED** [review]
Very good documentation changes, but please see my comments.Especially that one about hierarchy and Field\SubTypeSwitch class.

**Pull request merged**